### PR TITLE
Fix for concurrency bug 

### PR DESCRIPTION
**Pull request by miclegr**
## Description
When adding a point to an index without providing an id, we rely on `TypedIndex::currentLabel` to get one. But we are not doing it in a threadsafe way. This PR fixes that, solves #65 

### C++

`TypedIndex::currentLabel` is now `std::atomic` and it's post-incremented atomically to get a new id when needed

## Checklist

- [x] My code follows the code style of this project.
- [x] I have added and/or updated appropriate documentation (if applicable).
- [x] All new and existing tests pass locally with these changes.
- [x] I have run static code analysis (if available) and resolved any issues.
- [x] I have considered backward compatibility (if applicable).
- [x] I have confirmed that this PR does not introduce any security vulnerabilities.

### CODE REVIEW

**Comment by miclegr** [comment]
C++ linting failing due to [this](https://github.com/spotify/voyager/pull/69#issuecomment-2211289219)

File: cpp/TypedIndex.h [hunk 1]
```diff
@@ -360,7 +360,7 @@ class TypedIndex : public Index {
 
     int start = 0;
     if (!ep_added) {
-      size_t id = ids.size() ? ids.at(0) : (currentLabel);
+      size_t id = ids.size() ? ids.at(0) : (currentLabel.fetch_add(1));
```
Comment (markkohdev) [inline cpp/TypedIndex.h:363]: Is `.fetch_add(1)` actually what we want throughout this change?  Wouldn't that result in us incrementing `currentLabel` multiple times where we actually only wanted to increment it once?

**Review by markkohdev — CHANGES_REQUESTED** [review]
Left a question about the usage of the atomic.  I'm not sure this will result in behavioral parity with the existing implementation

File: cpp/TypedIndex.h (hunk 1 above)
Comment (miclegr) [reply cpp/TypedIndex.h]: Correct, I'm reverting this

Comment (miclegr) [reply cpp/TypedIndex.h:363]: Actually not, my original implementation is correct. Basically here we assign an id, then here:

**Review by markkohdev — APPROVED** [review]
👍 Thanks for addressing the comment!  This looks good to me, albeit race conditions are always a little tricky to verify fixes for

**Pull request merged**